File: controller/src/reconciler/namespace_ownership.rs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

//! Namespace claim v1. Namespace annotations are cluster-admin/controller
//! authority, not tenant-supplied labels. See docs/how-to/namespace-ownership.md.
//! Never read Secrets to establish ownership.

use k8s_openapi::{
    api::{apps::v1::Deployment, core::v1::Namespace},
    apimachinery::pkg::apis::meta::v1::ObjectMeta,
};
use kube::{
    Api, Client, ResourceExt,
    api::{DeleteParams, ListParams, Patch, PatchParams, PostParams, Preconditions},
    runtime::reflector::ObjectRef,
};
use serde_json::{Value, json};
use std::sync::LazyLock;
use tokio::sync::{Mutex, MutexGuard};

use crate::{crd::KarsSandbox, field_managers::CLAWSANDBOX};

pub const FINALIZER: &str = "kars.azure.com/namespace-cleanup";
pub const VERSION: &str = "kars.azure.com/namespace-claim-version";
pub const SOURCE_NAMESPACE: &str = "kars.azure.com/sandbox-namespace";
pub const SOURCE_NAME: &str = "kars.azure.com/sandbox-name";
pub const SOURCE_UID: &str = "kars.azure.com/sandbox-uid";
pub const NAMESPACE_UID: &str = "kars.azure.com/namespace-uid";
pub const PRESTAGE: &str = "kars.azure.com/namespace-prestage";

#[derive(Debug, thiserror::Error)]
pub enum Error {
    #[error("Kubernetes namespace ownership API error: {0}")]
    Kube(#[from] kube::Error),
    #[error("NamespaceOwnershipConflict: {0}; see docs/how-to/namespace-ownership.md")]
    Conflict(String),
    #[error("Namespace ownership serialization error: {0}")]
    Json(#[from] serde_json::Error),
}

fn conflict(message: &str) -> Error {
    Error::Conflict(message.into())
}

// Same-named CRs have different controller queue keys. Serialize their entire
// reconcile (including cleanup), not only the claim operation. Bounded stripes
// avoid an unbounded map of tenant-controlled names.
static LOCKS: LazyLock<Vec<Mutex<()>>> =
    LazyLock::new(|| (0..64).map(|_| Mutex::new(())).collect());

pub async fn lock(name: &str) -> MutexGuard<'static, ()> {
    let stripe = name.bytes().fold(0usize, |sum, byte| sum + byte as usize) % LOCKS.len();
    LOCKS[stripe].lock().await
}

/// Watch routing is only a hint; ensure() always re-reads the authoritative CR
/// and namespace. Neither labels nor watch payloads authorize a write.
pub fn to_sandbox_ref(namespace: Namespace) -> Option<ObjectRef<KarsSandbox>> {
    let meta = &namespace.metadata;
    let name = annotation(meta, SOURCE_NAME)?;
    let source_namespace = annotation(meta, SOURCE_NAMESPACE)?;
    if annotation(meta, VERSION) != Some("v1")
        || source_namespace.is_empty()
        || name.is_empty()
        || namespace.name_any() != format!("kars-{name}")
    {
        return None;
    }
    Some(ObjectRef::new(name).within(source_namespace))
}

fn annotation<'a>(meta: &'a ObjectMeta, key: &str) -> Option<&'a str> {
    meta.annotations.as_ref()?.get(key).map(String::as_str)
}

fn identity(meta: &ObjectMeta) -> Result<(&str, &str), Error> {
    match (meta.uid.as_deref(), meta.resource_version.as_deref()) {
        (Some(uid), Some(rv)) if !uid.is_empty() && !rv.is_empty() => Ok((uid, rv)),
        _ => Err(conflict(
            "object is missing its API-server UID/resourceVersion",
        )),
    }
}

fn owner_annotations(sandbox: &KarsSandbox) -> Value {
    json!({
        VERSION: "v1",
        SOURCE_NAMESPACE: sandbox.namespace(),
        SOURCE_NAME: sandbox.name_any(),
        SOURCE_UID: sandbox.metadata.uid,
    })
}

/// Only metadata is written when adopting; existing labels, workloads, Secrets,
/// and pod templates are never modified by this module.
fn claim_patch(namespace: &Namespace, sandbox: &KarsSandbox) -> Result<Value, Error> {
    let (uid, rv) = identity(&namespace.metadata)?;
    let mut annotations = owner_annotations(sandbox);
    annotations[PRESTAGE] = Value::Null;
    Ok(json!({"metadata": {
        "uid": uid, "resourceVersion": rv, "annotations": annotations,
    }}))
}

fn has_foreign_owner(meta: &ObjectMeta) -> bool {
    meta.owner_references
        .as_ref()
        .is_some_and(|refs| !refs.is_empty())
}

/// `true` means already claimed; `false` means legacy evidence is needed.
pub fn claimed(namespace: &Namespace, sandbox: &KarsSandbox) -> Result<bool, Error> {
    let (ns_uid, _) = identity(&namespace.metadata)?;
    identity(&sandbox.metadata)?;
    if namespace.name_any() != format!("kars-{}", sandbox.name_any())
        || has_foreign_owner(&namespace.metadata)
    {
        return Err(conflict("namespace name or ownerReferences do not match"));
    }
    if let Some(bound_uid) = annotation(&sandbox.metadata, NAMESPACE_UID)
        && bound_uid != ns_uid
    {
        return Err(conflict(
            "namespace was replaced; recorded namespace UID differs",
        ));
    }
    let meta = &namespace.metadata;
    let has_claim = [VERSION, SOURCE_NAMESPACE, SOURCE_NAME, SOURCE_UID, PRESTAGE]
        .iter()
        .any(|key| annotation(meta, key).is_some());
    if !has_claim {
        return Ok(false);
    }
    if annotation(meta, VERSION) != Some("v1")
        || annotation(meta, SOURCE_NAMESPACE) != sandbox.metadata.namespace.as_deref()
        || annotation(meta, SOURCE_NAME) != sandbox.metadata.name.as_deref()
    {
        return Err(conflict(
            "namespace is reserved for a different sandbox/workspace",
        ));
    }
    if annotation(meta, SOURCE_UID) != sandbox.metadata.uid.as_deref()
        || annotation(meta, PRESTAGE).is_some()
    {
        return Err(conflict("namespace is not bound to this Sandbox UID"));
    }
    Ok(true)
}

fn prestaged(namespace: &Namespace, sandbox: &KarsSandbox) -> bool {
    let meta = &namespace.metadata;
    namespace.name_any() == format!("kars-{}", sandbox.name_any())
        && annotation(meta, VERSION) == Some("v1")
        && annotation(meta, SOURCE_NAMESPACE) == sandbox.metadata.namespace.as_deref()
        && annotation(meta, SOURCE_NAME) == sandbox.metadata.name.as_deref()
        && annotation(meta, SOURCE_UID).is_none()
        && annotation(meta, PRESTAGE) == Some("bind-next-sandbox")
        && annotation(&sandbox.metadata, NAMESPACE_UID) == meta.uid.as_deref()
        && !has_foreign_owner(meta)
        && meta
            .creation_timestamp
            .as_ref()
            .zip(sandbox.metadata.creation_timestamp.as_ref())
            .is_some_and(|(ns_time, sandbox_time)| ns_time <= sandbox_time)
}

fn applied_field(meta: &ObjectMeta, path: &[&str]) -> bool {
    meta.managed_fields.as_ref().is_some_and(|fields| {
        fields.iter().any(|entry| {
            entry.manager.as_deref() == Some(CLAWSANDBOX)
                && entry.operation.as_deref() == Some("Apply")
                && entry.subresource.as_deref().is_none_or(str::is_empty)
                && entry.fields_v1.as_ref().is_some_and(|fields| {
                    let mut value = &fields.0;
                    for segment in path {
                        let Some(next) = value.get(*segment) else {
                            return false;
                        };
                        value = next;
                    }
                    true
                })
        })
    })
}

/// Legacy proof is conjunctive: current CR status + controller-managed namespace
/// and Deployment + Deployment creation strictly after this CR incarnation.
/// Equal timestamps are ambiguous (Kubernetes timestamps have second precision).
/// Namespace creation may precede the CR: the legacy CLI prestages credentials.
pub fn legacy_proof(namespace: &Namespace, deployment: &Deployment, sandbox: &KarsSandbox) -> bool {
    let name = sandbox.name_any();
    let ns_name = format!("kars-{name}");
    let label = |meta: &ObjectMeta, key: &str, value: &str| {
        meta.labels
            .as_ref()
            .and_then(|labels| labels.get(key))
            .is_some_and(|v| v == value)
    };
    sandbox
        .status
        .as_ref()
        .and_then(|status| status.namespace.as_deref())
        == Some(&ns_name)
        && sandbox
            .metadata
            .finalizers
            .as_ref()
            .is_some_and(|fs| fs.iter().any(|f| f == FINALIZER))
        && label(&namespace.metadata, "kars.azure.com/sandbox", &name)
        && label(&namespace.metadata, "kars.azure.com/role", "sandbox")
        && applied_field(
            &namespace.metadata,
            &["f:metadata", "f:labels", "f:kars.azure.com/sandbox"],
        )
        && deployment.name_any() == name
        && deployment.namespace().as_deref() == Some(&ns_name)
        && !has_foreign_owner(&deployment.metadata)
        && deployment.metadata.deletion_timestamp.is_none()
        && label(&deployment.metadata, "kars.azure.com/sandbox", &name)
        && deployment
            .metadata
            .labels
            .as_ref()
            .and_then(|labels| labels.get("kars.azure.com/parent-namespace"))
            .is_none_or(|parent| Some(parent.as_str()) == sandbox.metadata.namespace.as_deref())
        && applied_field(&deployment.metadata, &["f:spec"])
        && deployment
            .metadata
            .creation_timestamp
            .as_ref()
            .zip(sandbox.metadata.creation_timestamp.as_ref())
            .is_some_and(|(deployment_time, sandbox_time)| deployment_time > sandbox_time)
        && deployment.spec.as_ref().is_some_and(|spec| {
            spec.selector
                .match_labels
                .as_ref()
                .and_then(|labels| labels.get("kars.azure.com/sandbox"))
                == Some(&name)
                && spec
                    .template
                    .metadata
                    .as_ref()
                    .is_some_and(|meta| label(meta, "kars.azure.com/sandbox", &name))
        })
}

async fn unique_sandbox(client: &Client, sandbox: &KarsSandbox) -> Result<(), Error> {
    let api: Api<KarsSandbox> = Api::all(client.clone());
    let list = api
        .list(&ListParams::default().fields(&format!("metadata.name={}", sandbox.name_any())))
        .await?;
    if list.items.len() != 1
        || list.items[0].metadata.uid != sandbox.metadata.uid
        || list.items[0].metadata.namespace != sandbox.metadata.namespace
    {
        return Err(conflict(
            "sandbox name is not unique across workspaces; no namespace was claimed",
        ));
    }
    Ok(())
}

async fn live_sandbox(client: &Client, sandbox: &KarsSandbox) -> Result<KarsSandbox, Error> {
    identity(&sandbox.metadata)?;
    let ns = sandbox
        .namespace()
        .filter(|ns| !ns.is_empty())
        .ok_or_else(|| conflict("Sandbox source namespace is missing"))?;
    let api: Api<KarsSandbox> = Api::namespaced(client.clone(), &ns);
    let live = api.get(&sandbox.name_any()).await?;
    if live.metadata.uid != sandbox.metadata.uid {
        return Err(conflict("Sandbox was recreated during reconciliation"));
    }
    Ok(live)
}

async fn patch_sandbox(
    client: &Client,
    sandbox: &KarsSandbox,
    mut metadata: Value,
) -> Result<KarsSandbox, Error> {
    let (uid, rv) = identity(&sandbox.metadata)?;
    metadata["uid"] = json!(uid);
    metadata["resourceVersion"] = json!(rv);
    let api: Api<KarsSandbox> = Api::namespaced(client.clone(), &sandbox.namespace().unwrap());
    Ok(api
        .patch(
            &sandbox.name_any(),
            &PatchParams::default(),
            &Patch::Merge(json!({"metadata": metadata})),
        )
        .await?)
}

fn new_namespace(sandbox: &KarsSandbox) -> Result<Namespace, Error> {
    Ok(serde_json::from_value(json!({
        "apiVersion": "v1", "kind": "Namespace",
        "metadata": {
            "name": format!("kars-{}", sandbox.name_any()),
            "annotations": owner_annotations(sandbox),
            "labels": {
                "app.kubernetes.io/name": "kars",
                "app.kubernetes.io/component": "sandbox",
                "kars.azure.com/sandbox": sandbox.name_any(),
                "kars.azure.com/role": "sandbox",
                "kars.azure.com/isolated": "strict",
                "pod-security.kubernetes.io/enforce": "privileged",
                "pod-security.kubernetes.io/audit": "baseline",
                "pod-security.kubernetes.io/warn": "baseline"
            }
        }
    }))?)
}

/// Establish authority before *any* target-namespace operations. HTTP conflicts
/// propagate to the controller's retry queue; every retry re-reads all evidence.
pub async fn ensure(
    client: &Client,
    observed: &KarsSandbox,
) -> Result<(KarsSandbox, Option<Namespace>), Error> {
    let mut sandbox = live_sandbox(client, observed).await?;
    let api: Api<Namespace> = Api::all(client.clone());
    let name = format!("kars-{}", sandbox.name_any());
    let mut namespace = api.get_opt(&name).await?;
    if namespace.is_none() {
        unique_sandbox(client, &sandbox).await?;
        if sandbox.metadata.deletion_timestamp.is_some() {
            return Ok((sandbox, None));
        }
        if annotation(&sandbox.metadata, NAMESPACE_UID).is_some() {
            return Err(conflict(
                "bound namespace is missing; explicit operator recovery is required",
            ));
        }
    }
    if let Some(ns) = &namespace {
        if ns.metadata.deletion_timestamp.is_some() && sandbox.metadata.deletion_timestamp.is_none()
        {
            return Err(conflict(
                "namespace is terminating; workload reconciliation is stopped",
            ));
        }
        if !prestaged(ns, &sandbox) && !claimed(ns, &sandbox)? {
            unique_sandbox(client, &sandbox).await?;
            let deployments: Api<Deployment> = Api::namespaced(client.clone(), &name);
            let deployment = deployments.get_opt(&sandbox.name_any()).await?;
            if !deployment
                .as_ref()
                .is_some_and(|d| legacy_proof(ns, d, &sandbox))
            {
                return Err(conflict(
                    "legacy namespace ownership is unproven; preserve it and run upgrade preflight",
                ));
            }
        } else if prestaged(ns, &sandbox) {
            unique_sandbox(client, &sandbox).await?;
        }
    }
    if sandbox.metadata.deletion_timestamp.is_none()
        && !sandbox
            .metadata
            .finalizers
            .as_ref()
            .is_some_and(|fs| fs.iter().any(|f| f == FINALIZER))
    {
        let mut finalizers = sandbox.metadata.finalizers.clone().unwrap_or_default();
        finalizers.push(FINALIZER.into());
        sandbox = patch_sandbox(client, &sandbox, json!({"finalizers": finalizers})).await?;
    }
    if namespace.is_none() {
        match api
            .create(&PostParams::default(), &new_namespace(&sandbox)?)
            .await
        {
            Ok(created) => namespace = Some(created),
            Err(kube::Error::Api(error)) if error.code == 409 => {
                // A concurrent creator won. Never turn this into an apply/adopt.
                let existing = api.get(&name).await?;
                if !claimed(&existing, &sandbox)? {
                    return Err(conflict("namespace creation raced an unclaimed namespace"));
                }
                namespace = Some(existing);
            }
            Err(error) => return Err(error.into()),
        }
    }
    let mut namespace = namespace.unwrap();
    if prestaged(&namespace, &sandbox) || !claimed(&namespace, &sandbox)? {
        namespace = api
            .patch(
                &name,
                &PatchParams::default(),
                &Patch::Merge(claim_patch(&namespace, &sandbox)?),
            )
            .await?;
    }
    // Re-read the CR before persisting the backlink. If its UID or deletion
    // state changed, do not proceed into any workload operations.
    let live = live_sandbox(client, &sandbox).await?;
    if live.metadata.deletion_timestamp != sandbox.metadata.deletion_timestamp {
        return Err(conflict("Sandbox deletion started during namespace claim"));
    }
    sandbox = live;
    if annotation(&sandbox.metadata, NAMESPACE_UID)
        .is_some_and(|uid| Some(uid) != namespace.metadata.uid.as_deref())
    {
        return Err(conflict("Sandbox namespace binding changed during claim"));
    }
    if annotation(&sandbox.metadata, NAMESPACE_UID) != namespace.metadata.uid.as_deref() {
        sandbox = patch_sandbox(
            client,
            &sandbox,
            json!({"annotations": {NAMESPACE_UID: namespace.metadata.uid}}),
        )
        .await?;
    }
    recheck(client, &sandbox, &namespace).await?;
    Ok((sandbox, Some(namespace)))
}

pub async fn recheck(
    client: &Client,
    sandbox: &KarsSandbox,
    expected: &Namespace,
) -> Result<Namespace, Error> {
    let api: Api<Namespace> = Api::all(client.clone());
    let live = api.get(&expected.name_any()).await?;
    if live.metadata.uid != expected.metadata.uid || !claimed(&live, sandbox)? {
        return Err(conflict(
            "namespace ownership changed during reconciliation",
        ));
    }
    if live.metadata.deletion_timestamp.is_some() && sandbox.metadata.deletion_timestamp.is_none() {
        return Err(conflict("namespace deletion started during reconciliation"));
    }
    Ok(live)
}

/// Read-only gate for other controllers consuming sandbox-local resources.
/// A missing namespace is not authority to read or create anything there.
/// Legacy namespaces wait for the Sandbox reconciler to establish their claim.
pub async fn verify_target(
    client: &Client,
    source_namespace: &str,
    name: &str,
) -> Result<bool, Error> {
    let valid_label = |value: &str| {
        !value.is_empty()
            && value.len() <= 63
            && value
                .bytes()
                .all(|b| b.is_ascii_lowercase() || b.is_ascii_digit() || b == b'-')
            && value.as_bytes()[0].is_ascii_alphanumeric()
            && value.as_bytes()[value.len() - 1].is_ascii_alphanumeric()
    };
    if !valid_label(source_namespace) || !valid_label(name) {
        return Err(conflict(
            "target name/workspace is not a Kubernetes DNS label",
        ));
    }
    let namespaces: Api<Namespace> = Api::all(client.clone());
    let Some(namespace) = namespaces.get_opt(&format!("kars-{name}")).await? else {
        return Ok(false);
    };
    let sandboxes: Api<KarsSandbox> = Api::namespaced(client.clone(), source_namespace);
    let sandbox = sandboxes
        .get_opt(name)
        .await?
        .ok_or_else(|| conflict("target namespace has no Sandbox owner in this workspace"))?;
    if sandbox.name_any() != name || sandbox.namespace().as_deref() != Some(source_namespace) {
        return Err(conflict(
            "target Sandbox identity differs from the requested workspace/name",
        ));
    }
    if !claimed(&namespace, &sandbox)? {
        return Err(conflict(
            "target namespace ownership has not been established",
        ));
    }
    Ok(true)
}

/// Accepted deletion is sufficient; namespace claims prevent reuse while it
/// terminates. Waiting for disappearance would deadlock a CR inside its own
/// runtime namespace, whose deletion itself waits for this CR's finalizer.
pub async fn delete(
    client: &Client,
    sandbox: &KarsSandbox,
    namespace: Option<&Namespace>,
) -> Result<(), Error> {
    let Some(namespace) = namespace else {
        return Ok(());
    };
    let live = match recheck(client, sandbox, namespace).await {
        Err(Error::Kube(kube::Error::Api(error))) if error.code == 404 => return Ok(()),
        result => result?,
    };
    let (uid, rv) = identity(&live.metadata)?;
    let api: Api<Namespace> = Api::all(client.clone());
    if live.metadata.deletion_timestamp.is_none() {
        match api
            .delete(
                &live.name_any(),
                &DeleteParams {
                    preconditions: Some(Preconditions {
                        uid: Some(uid.into()),
                        resource_version: Some(rv.into()),
                    }),
                    ..Default::default()
                },
            )
            .await
        {
            Ok(_) => {}
            Err(kube::Error::Api(error)) if error.code == 404 => {}
            Err(error) => return Err(error.into()),
        }
    }
    Ok(())
}

pub async fn remove_finalizer(client: &Client, sandbox: &KarsSandbox) -> Result<(), Error> {
    let sandbox = live_sandbox(client, sandbox).await?;
    let finalizers: Vec<_> = sandbox
        .metadata
        .finalizers
        .iter()
        .flatten()
        .filter(|f| f.as_str() != FINALIZER)
        .cloned()
        .collect();
    patch_sandbox(client, &sandbox, json!({"finalizers": finalizers})).await?;
    Ok(())
}

pub async fn report_conflict(
    client: &Client,
    sandbox: &KarsSandbox,
    message: &str,
) -> Result<(), Error> {
    let sandbox = live_sandbox(client, sandbox).await?;
    let (uid, rv) = identity(&sandbox.metadata)?;
    let mut patch =
        crate::status::build_degraded_status_patch(&sandbox, "NamespaceOwnershipConflict", message);
    let current_status = serde_json::to_value(&sandbox.status)?;
    if patch["status"].as_object().is_some_and(|fields| {
        fields
            .iter()
            .all(|(key, value)| current_status.get(key) == Some(value))
    }) {
        return Ok(());
    }
    patch["metadata"] = json!({"uid": uid, "resourceVersion": rv});
    let api: Api<KarsSandbox> = Api::namespaced(client.clone(), &sandbox.namespace().unwrap());
    api.patch_status(
        &sandbox.name_any(),
        &PatchParams::default(),
        &Patch::Merge(patch),
    )
    .await?;
    Ok(())
}

#[cfg(test)]
#[path = "namespace_ownership_tests.rs"]
mod tests;

File: cli/src/lib/namespace-ownership.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

type Execute = (
  file: string, args: readonly string[], options: { stdio: "pipe"; input?: string },
) => Promise<{ stdout: string }>;

export const CLAIM = {
  version: "kars.azure.com/namespace-claim-version",
  namespace: "kars.azure.com/sandbox-namespace",
  name: "kars.azure.com/sandbox-name",
  uid: "kars.azure.com/sandbox-uid",
  namespaceUid: "kars.azure.com/namespace-uid",
  prestage: "kars.azure.com/namespace-prestage",
} as const;
const FINALIZER = "kars.azure.com/namespace-cleanup";
const MANAGER = "kars-controller/karssandbox";
const GUIDE = "docs/how-to/namespace-ownership.md";

export interface OwnershipObject {
  apiVersion?: string;
  kind?: string;
  metadata: {
    name?: string; namespace?: string; uid?: string; resourceVersion?: string;
    creationTimestamp?: string; deletionTimestamp?: string;
    annotations?: Record<string, string>; labels?: Record<string, string>;
    ownerReferences?: unknown[]; finalizers?: string[];
    managedFields?: Array<{
      manager?: string; operation?: string; subresource?: string;
      fieldsV1?: Record<string, unknown>;
    }>;
  };
  status?: { namespace?: string };
  spec?: {
    selector?: { matchLabels?: Record<string, string> };
    template?: { metadata?: { labels?: Record<string, string> } };
  };
}

function fail(message: string): never {
  throw new Error(`NamespaceOwnershipConflict: ${message}; see ${GUIDE}`);
}

function identity(object: OwnershipObject): void {
  if (!object.metadata?.uid || !object.metadata.resourceVersion || !object.metadata.name) {
    fail("API object omitted name/UID/resourceVersion");
  }
}

function validateTarget(name: string, namespace: string): void {
  for (const value of [name, namespace]) {
    if (!value || value.length > 63 || !/^[a-z0-9](?:[a-z0-9-]*[a-z0-9])?$/.test(value)) {
      fail("target name/workspace is not a Kubernetes DNS label");
    }
  }
}

export function claimAnnotations(sandbox: OwnershipObject): Record<string, string> {
  identity(sandbox);
  if (!sandbox.metadata.namespace || !sandbox.metadata.name) fail("Sandbox identity is incomplete");
  return {
    [CLAIM.version]: "v1", [CLAIM.namespace]: sandbox.metadata.namespace,
    [CLAIM.name]: sandbox.metadata.name, [CLAIM.uid]: sandbox.metadata.uid!,
  };
}

/** Strict checks shared by diagnostics, explicit adoption, and credential prestaging. */
export function namespaceClaimed(namespace: OwnershipObject, sandbox: OwnershipObject): boolean {
  identity(namespace);
  identity(sandbox);
  if (namespace.metadata.name !== `kars-${sandbox.metadata.name}`
      || namespace.metadata.ownerReferences?.length) fail("namespace name or ownerReferences do not match");
  const bound = sandbox.metadata.annotations?.[CLAIM.namespaceUid];
  if (bound !== undefined && bound !== namespace.metadata.uid) fail("namespace was replaced");
  const annotations = namespace.metadata.annotations ?? {};
  const keys = [CLAIM.version, CLAIM.namespace, CLAIM.name, CLAIM.uid, CLAIM.prestage];
  if (!keys.some(key => annotations[key] !== undefined)) return false;
  for (const [key, value] of Object.entries(claimAnnotations(sandbox))) {
    if (annotations[key] !== value) fail("namespace is reserved for a different Sandbox UID/workspace");
  }
  if (annotations[CLAIM.prestage] !== undefined) fail("namespace is not bound to this Sandbox UID");
  return true;
}

export function namespacePrestaged(namespace: OwnershipObject, sandbox: OwnershipObject): boolean {
  const annotations = namespace.metadata.annotations ?? {};
  const nsTime = Date.parse(namespace.metadata.creationTimestamp ?? "");
  const sandboxTime = Date.parse(sandbox.metadata.creationTimestamp ?? "");
  return namespace.metadata.name === `kars-${sandbox.metadata.name}`
    && !namespace.metadata.ownerReferences?.length
    && annotations[CLAIM.version] === "v1"
    && annotations[CLAIM.namespace] === sandbox.metadata.namespace
    && annotations[CLAIM.name] === sandbox.metadata.name
    && annotations[CLAIM.uid] === undefined
    && annotations[CLAIM.prestage] === "bind-next-sandbox"
    && !!namespace.metadata.uid
    && sandbox.metadata.annotations?.[CLAIM.namespaceUid] === namespace.metadata.uid
    && Number.isFinite(nsTime) && Number.isFinite(sandboxTime) && nsTime <= sandboxTime;
}

function appliedField(object: OwnershipObject, path: string[]): boolean {
  return object.metadata.managedFields?.some(entry => {
    if (entry.manager !== MANAGER || entry.operation !== "Apply" || entry.subresource) return false;
    let value: unknown = entry.fieldsV1;
    for (const key of path) {
      if (!value || typeof value !== "object" || !(key in value)) return false;
      value = (value as Record<string, unknown>)[key];
    }
    return true;
  }) ?? false;
}

/** Keep this conjunctive test in sync with controller namespace_ownership::legacy_proof. */
export function legacyNamespaceProof(
  namespace: OwnershipObject, deployment: OwnershipObject, sandbox: OwnershipObject,
): boolean {
  const name = sandbox.metadata.name;
  const target = `kars-${name}`;
  const parent = deployment.metadata.labels?.["kars.azure.com/parent-namespace"];
  const deployed = Date.parse(deployment.metadata.creationTimestamp ?? "");
  const created = Date.parse(sandbox.metadata.creationTimestamp ?? "");
  return sandbox.status?.namespace === target
    && (sandbox.metadata.finalizers?.includes(FINALIZER) ?? false)
    && namespace.metadata.labels?.["kars.azure.com/sandbox"] === name
    && namespace.metadata.labels?.["kars.azure.com/role"] === "sandbox"
    && appliedField(namespace, ["f:metadata", "f:labels", "f:kars.azure.com/sandbox"])
    && deployment.metadata.name === name && deployment.metadata.namespace === target
    && !deployment.metadata.ownerReferences?.length && !deployment.metadata.deletionTimestamp
    && deployment.metadata.labels?.["kars.azure.com/sandbox"] === name
    && (parent === undefined || parent === sandbox.metadata.namespace)
    && appliedField(deployment, ["f:spec"])
    && Number.isFinite(deployed) && Number.isFinite(created) && deployed > created
    && deployment.spec?.selector?.matchLabels?.["kars.azure.com/sandbox"] === name
    && deployment.spec?.template?.metadata?.labels?.["kars.azure.com/sandbox"] === name;
}

async function get(execute: Execute, kind: string, name: string, namespace?: string): Promise<OwnershipObject | undefined> {
  // --ignore-not-found suppresses only 404. Auth, transport, and parsing errors
  // must stop the operation; they are never interpreted as an empty cluster.
  const { stdout } = await execute("kubectl", [
    "get", kind, name, ...(namespace ? ["-n", namespace] : []),
    "--ignore-not-found", "--show-managed-fields=true", "-o", "json",
  ], { stdio: "pipe" });
  if (!String(stdout).trim()) return undefined;
  const object = JSON.parse(String(stdout)) as OwnershipObject;
  identity(object);
  return object;
}

async function sandboxes(execute: Execute): Promise<OwnershipObject[]> {
  const { stdout } = await execute("kubectl", [
    "get", "karssandboxes", "-A", "--show-managed-fields=true", "-o", "json",
  ], { stdio: "pipe" });
  const list = JSON.parse(String(stdout)) as { items?: OwnershipObject[] };
  if (!Array.isArray(list.items)) fail("Sandbox inventory is incomplete");
  for (const sandbox of list.items) {
    identity(sandbox);
    if (!sandbox.metadata.namespace) fail("Sandbox inventory omitted its workspace namespace");
  }
  return list.items;
}

export async function inspectNamespaceOwnership(execute: Execute): Promise<string[]> {
  const inventory = await sandboxes(execute);
  const diagnostics: string[] = [];
  for (const sandbox of inventory) {
    const name = sandbox.metadata.name!;
    const label = `${sandbox.metadata.namespace}/${name}`;
    const target = await get(execute, "namespace", `kars-${name}`);
    const unique = inventory.filter(s => s.metadata.name === name).length === 1;
    if (!target) {
      if (!unique) fail(`${label}: same-name Sandboxes exist in different workspaces`);
      if (sandbox.metadata.annotations?.[CLAIM.namespaceUid] && !sandbox.metadata.deletionTimestamp) {
        fail(`${label}: previously bound namespace is missing`);
      }
      diagnostics.push(`${label}: ${sandbox.metadata.deletionTimestamp ? "cleanup pending" : "new namespace (atomic create)"}`);
      continue;
    }
    if (target.metadata.deletionTimestamp && !sandbox.metadata.deletionTimestamp) fail(`${label}: namespace is terminating`);
    if (namespacePrestaged(target, sandbox)) {
      if (!unique) fail(`${label}: prestage target is ambiguous`);
      diagnostics.push(`${label}: explicit prestage (will bind UID)`);
      continue;
    }
    if (namespaceClaimed(target, sandbox)) {
      diagnostics.push(`${label}: namespace claim v1 verified`);
      continue;
    }
    if (!unique) fail(`${label}: legacy namespace has multiple possible owners`);
    const deployment = await get(execute, "deployment", name, `kars-${name}`);
    if (!deployment || !legacyNamespaceProof(target, deployment, sandbox)) {
      fail(`${label}: legacy ownership is unproven; resources preserved; explicit adoption required`);
    }
    diagnostics.push(`${label}: unambiguous legacy deployment (metadata-only adoption)`);
  }
  return diagnostics;
}

/** Explicit administrator decision, never a preflight side effect or a force takeover. */
export async function adoptNamespace(
  execute: Execute, name: string, sourceNamespace: string, sandboxUid: string, namespaceUid: string,
): Promise<void> {
  validateTarget(name, sourceNamespace);
  const inventory = await sandboxes(execute);
  const candidates = inventory.filter(s => s.metadata.name === name);
  const sandbox = candidates[0];
  if (candidates.length !== 1 || sandbox.metadata.namespace !== sourceNamespace
      || sandbox.metadata.uid !== sandboxUid || sandbox.metadata.deletionTimestamp) {
    fail("reviewed Sandbox UID/workspace is no longer the unique live owner");
  }
  const namespace = await get(execute, "namespace", `kars-${name}`);
  if (!namespace || namespace.metadata.uid !== namespaceUid || namespace.metadata.deletionTimestamp) {
    fail("reviewed namespace UID is missing, replaced, or terminating");
  }
  // A partial/foreign claim is not an invitation to overwrite it.
  namespaceClaimed(namespace, sandbox);
  await execute("kubectl", ["patch", "namespace", `kars-${name}`, "--type=merge", "-p",
    JSON.stringify({ metadata: {
      uid: namespaceUid, resourceVersion: namespace.metadata.resourceVersion,
      annotations: claimAnnotations(sandbox),
    } }),
  ], { stdio: "pipe" });
}

/** Preserve add's pre-CR credential staging, with an explicit two-way namespace
 * reservation. Returning the namespace UID lets add include it in the CR CREATE.
 * Interrupted explicit reservations can be resumed, but unmarked namespaces
 * are never claimed merely because their name matches. */
export async function prepareCredentialNamespace(
  execute: Execute, name: string, sourceNamespace: string,
): Promise<string> {
  validateTarget(name, sourceNamespace);
  const inventory = await sandboxes(execute);
  const candidates = inventory.filter(s => s.metadata.name === name);
  if (candidates.some(s => s.metadata.namespace !== sourceNamespace)) fail("sandbox name belongs to another workspace");
  const sandbox = candidates[0];
  const existing = await get(execute, "namespace", `kars-${name}`);
  if (existing) {
    const annotations = existing.metadata.annotations ?? {};
    if (!sandbox && !existing.metadata.deletionTimestamp && !existing.metadata.ownerReferences?.length
        && annotations[CLAIM.version] === "v1"
        && annotations[CLAIM.namespace] === sourceNamespace && annotations[CLAIM.name] === name
        && annotations[CLAIM.uid] === undefined && annotations[CLAIM.prestage] === "bind-next-sandbox") {
      return existing.metadata.uid!;
    }
    if (!sandbox || existing.metadata.deletionTimestamp) fail("existing namespace needs explicit adoption before credential staging");
    if (!namespaceClaimed(existing, sandbox)) {
      const deployment = await get(execute, "deployment", name, `kars-${name}`);
      if (!deployment || !legacyNamespaceProof(existing, deployment, sandbox)) fail("credential target ownership is unproven");
    }
    if (sandbox.metadata.deletionTimestamp) fail("Sandbox is terminating");
    return existing.metadata.uid!;
  }
  if (sandbox) fail("existing Sandbox has no namespace; wait for the controller");
  const { stdout } = await execute("kubectl", ["create", "-f", "-", "-o", "json"], {
    input: JSON.stringify({
      apiVersion: "v1", kind: "Namespace", metadata: {
        name: `kars-${name}`,
        annotations: {
          [CLAIM.version]: "v1", [CLAIM.namespace]: sourceNamespace,
          [CLAIM.name]: name, [CLAIM.prestage]: "bind-next-sandbox",
        },
        labels: {
          "app.kubernetes.io/name": "kars", "app.kubernetes.io/component": "sandbox",
          "kars.azure.com/sandbox": name, "kars.azure.com/role": "sandbox",
          "kars.azure.com/isolated": "strict",
          "pod-security.kubernetes.io/enforce": "privileged",
          "pod-security.kubernetes.io/audit": "baseline",
          "pod-security.kubernetes.io/warn": "baseline",
        },
      },
    }), stdio: "pipe",
  });
  const created = JSON.parse(String(stdout)) as OwnershipObject;
  identity(created);
  return created.metadata.uid!;
}
